modules/custom/openy_focal_point/src/Plugin/Field/FieldWidget/OpenYFocalPointImageWidget.php

```diff
@@ -0,0 +1,225 @@
+
+      // Search through $form_state['input'] for target_id => 'media:XX' element. If found
+      // use "entity_browser_widget" right next to it to get id that can be used
+      // to load image styles for the paragraph.
+      $paragraph_type = self::getParagraphInfo($fid, $form_state);
+
+      // Another big assumption here. We assume that Media has view modes named
+      // "prgf_<paragraph_name>".
+      $display = \Drupal::entityTypeManager()->getStorage('entity_view_display')->load('media.image.prgf_' . $paragraph_type);
+
+      if ($display) {
+        $components = $display->getComponents();
+        // We assume that view mode displays only single field -- image.
+        $image_component = reset($components);
+
+        $used_breakpoints = [
+          $image_component['settings']['image_style'],
+        ];
+        // This is to pull mobile breakpoints from Blazy formatter. This is
+        // not being used at the moment. Keeping it for the future reference.
+//        foreach ($image_component['settings']['breakpoints'] as $breakpoint) {
+//          if (!empty($breakpoint['image_style'])) {
+//            $used_breakpoints[] = $breakpoint['image_style'];
+//          }
+//        }
+
+        $element['preview']['preview_link'] = [
+          '#type' => 'inline_template',
+          '#template' => '{{ focal_point_link }}  |  {{ crop_image_link }}',
+          '#context' => [
+            'focal_point_link' => self::createFocalPointEditLink($fid, $element['#field_name'], $element_selectors, $default_focal_point_value, $used_breakpoints),
+            'crop_image_link' => self::createCropLink($fid, $element['#field_name'], $element_selectors, $default_focal_point_value, $used_breakpoints),
+          ],
+        ];
+
+        $element['preview']['preview_link_note'] = [
+          '#type' => 'inline_template',
+          '#template' => '<div class="focal-point-preview-link-note">{{ note }}</div>',
+          '#context' => [
+            'note' => new TranslatableMarkup('Note: Focal Point and Crop Image cannot both be applied to the same image.'),
+          ]
+        ];
+      }
+
+    }
+
+    return $element;
+  }
+
+  /**
+   * Gets paragraph info data. We do a few assumptions here. We assume that
+   * there is a top level fields (references to paragraphs) like field_content,
+   * field_header_content etc. Also there should be subforms inside of the
+   * paragraphs. So our data is located in structure like:
+   * $form_state->getValues()['field_content'][0]['subform']['field_prgf_image']['entity_browser_widget_paragraph_info'];
+   *
+   * @see OpenYFocalPointEntityReferenceBrowserWidget::formElement() where we
+   * create a hidden field 'entity_browser_widget_paragraph_info'.
+   */
+  protected static function getParagraphInfo($fid, FormStateInterface $form_state) {
+    $query = \Drupal::entityTypeManager()->getStorage('media')->getQuery();
```

Comment:
we should pass service via DI

Reply:
we use this service in public static method, so we don't have instance of the form. This is why we can not use DI here.